src/buffer/out/UTextAdapter.cpp

```diff
@@ -0,0 +1,328 @@
+    }
+
+    if (!utextAccess(ut, nativeStart, true))
+    {
+        return 0;
+    }
+
+    nativeLimit = std::min(ut->chunkNativeLimit, nativeLimit);
+
+    if (destCapacity <= 0)
+    {
+        return gsl::narrow_cast<int32_t>(nativeLimit - nativeStart);
+    }
+
+    const auto& textBuffer = *static_cast<const TextBuffer*>(ut->context);
+    const auto y = accessCurrentRow(ut);
+    const auto offset = ut->chunkNativeStart - nativeStart;
+    const auto text = textBuffer.GetRowByOffset(y).GetText().substr(gsl::narrow_cast<size_t>(std::max<int64_t>(0, offset)));
+    const auto destCapacitySizeT = gsl::narrow_cast<size_t>(destCapacity);
+    const auto length = std::min(destCapacitySizeT, text.size());
+
+    memcpy(dest, text.data(), length * sizeof(char16_t));
+
+    if (length < destCapacitySizeT)
+    {
+#pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+        dest[length] = 0;
+    }
+
+    return gsl::narrow_cast<int32_t>(length);
+}
+catch (...)
+{
+    // The only thing that can fail is GetRowByOffset() which in turn can only fail when VirtualAlloc() fails.
+    *status = U_MEMORY_ALLOCATION_ERROR;
+    return 0;
+}
+
+static constexpr UTextFuncs utextFuncs{
+    .tableSize = sizeof(UTextFuncs),
+    .clone = utextClone,
+    .nativeLength = utextNativeLength,
+    .access = utextAccess,
+    .extract = utextExtract,
+};
+
+// Creates a UText from the given TextBuffer that spans rows [rowBeg,RowEnd).
+UText UTextFromTextBuffer(const TextBuffer& textBuffer, til::CoordType rowBeg, til::CoordType rowEnd, UErrorCode* status) noexcept
+{
+    __assume(status != nullptr);
+
+    UText ut = UTEXT_INITIALIZER;
+    ut.providerProperties = (1 << UTEXT_PROVIDER_LENGTH_IS_EXPENSIVE) | (1 << UTEXT_PROVIDER_STABLE_CHUNKS);
+    ut.pFuncs = &utextFuncs;
+    ut.context = &textBuffer;
+    accessCurrentRow(&ut) = rowBeg - 1; // the utextAccess() below will advance this by 1.
+    accessRowRange(&ut) = { rowBeg, rowEnd };
+
+    utextAccess(&ut, 0, true);
+    return ut;
+}
+
+URegularExpression* CreateURegularExpression(const std::wstring_view& pattern, uint32_t flags, UErrorCode* status) noexcept
+{
+#pragma warning(suppress : 26490) // Don't use reinterpret_cast (type.1).
+    const auto re = uregex_open(reinterpret_cast<const char16_t*>(pattern.data()), gsl::narrow_cast<int32_t>(pattern.size()), flags, nullptr, status);
+    // ICU describes the time unit as being dependent on CPU performance and "typically [in] the order of milliseconds",
+    // but this claim seems highly outdated already. On my CPU from 2021, a limit of 4096 equals roughly 600ms.
+    uregex_setTimeLimit(re, 4096, status);
+    uregex_setStackLimit(re, 4 * 1024 * 1024, status);
+    return re;
+}
+
+// Returns an inclusive point range given a text start and end position.
+// This function is designed to be used with uregex_start64/uregex_end64.
+til::point_span BufferRangeFromMatch(UText* ut, URegularExpression* re)

```

Comment:
nit: this is somewhat generically-named for being in the global namespace. `BufferRangeFromURegularExpressionMatch`?